src/lib/OS/MemRegion.cc

```diff
@@ -1,67 +1,313 @@
+uint64_t MemRegion::getStackEnd() const { return stackEnd_; }
 
-uint16_t MemRegion::getHeapSize() const { return heapSize_; }
+size_t MemRegion::getStackSize() const { return stackSize_; }
 
-uint64_t MemRegion::getInitialStackStart() const { return initStackStart_; }
+uint64_t MemRegion::getInitialStackPtr() const { return initStackPtr_; }
 
-uint64_t MemRegion::getBrk() const { return brk_; }
+uint64_t MemRegion::getHeapStart() const { return heapStart_; }
+
+uint64_t MemRegion::getHeapEnd() const { return heapEnd_; }
 
-uint64_t MemRegion::getBrkStart() const { return startBrk_; }
+size_t MemRegion::getHeapSize() const { return heapSize_; }
+
+uint64_t MemRegion::getBrk() const { return brk_; }
 
 uint64_t MemRegion::getMmapStart() const { return mmapStart_; }
 
 uint64_t MemRegion::getMemSize() const { return memSize_; }
 
 uint64_t MemRegion::updateBrkRegion(uint64_t newBrk) {
-  if (newBrk > maxHeapAddr_) {
-    // This needs to fixed such that more extra memory allocation is mmapd.
-    std::cerr
-        << "Attemped to allocate more memory than is available to the process "
-        << std::endl;
+  if (newBrk < heapStart_) {
+    return brk_;
+  }
+  if (newBrk > heapEnd_) {
+    // TODO: This needs to fixed such that more extra memory allocation is
+    // mmapd.
+    std::cerr << "[SimEng:MemRegion] Attemped to allocate more memory on the "
+                 "heap than is available to the process. Please increase the "
+                 "{Process-Image:{Heap-Size: <size>}} parameter in YAML "
+                 "config file used to run this simulation."
+              << std::endl;
     std::exit(1);
   }
-  if (newBrk > startBrk_) {
+
+  if (newBrk > brk_) {
     brk_ = newBrk;
   }
   return brk_;
 }
 
-uint64_t MemRegion::mmapRegion(uint64_t addr, uint64_t length, int fd, int prot,
-                               int flags) {
-  // Need to deal with addr
-  VirtMemArea* new_vma = new VirtMemArea();
-  new_vma->length = roundUpMemAddr(length, pageSize_);
-  return vma_ll.addVma(new_vma, mmapStart_, pageSize_);
+uint64_t MemRegion::addVma(VMA* vma, uint64_t startAddr) {
+  VMA* curr = vm_head_;
+  size_t size = vma->vmSize_;
+  // When starAddr is not 0, search for an available address range
+  // that can hold the new VMA with a starting address that is greater than or
+  // equal to the specified startAddr. The following algorithm retrieves the
+  // last existing VMA object before the address range so that the new VMA can
+  // be linked between two existing VMAs. If no available address range is
+  // found, then the new VMA is allocated at the end of the VMA list.
+  if (startAddr && vm_size_ > 0) {
+    while (curr->vmNext_ != nullptr) {
+      if (curr->vmEnd_ <= startAddr &&
+          curr->vmNext_->vmStart_ >= (startAddr + size)) {
+        break;
+      }
+      curr = curr->vmNext_;
+    }
+  }
+
+  bool allocated = false;
+  // If Vma list has multiple VMAs then starting from curr (VMA) check if the
```

Comment:
If the VMA list has multiple VMAs, then ...

Reply:
done